### PR TITLE
Support the batch mode in load stage ( just for tikv)

### PR DESCRIPTION
**Pull request by liukun4515**
In order to bench the tikv in the batch mode, I add the batch interface in the go-ycsb:

- add the `BatchInsert` function in ycsb.DB interface
- add the `DoBatchInsert` function in the ycsb.workload interface

Now the batch mode is just supported in load stage, I will support it for run stage in the next pr.

### CODE REVIEW

**Comment by liukun4515** [comment]
PTAL @siddontang

File: db/basic/db.go [hunk 1]
```diff
@@ -209,6 +209,32 @@ func (db *basicDB) Insert(ctx context.Context, table string, key string, values
...
+	state := ctx.Value(stateKey).(*basicState)
+
+	db.delay(ctx, state)
+
+	if !db.verbose {
+		return nil
+	}
+	buf := state.buf
+	for i, key := range keys {
+		s := fmt.Sprintf("INSERT %s %s [ ", table, key)
```
Comment (Connor1996) [inline db/basic/db.go]: maybe you can extract common part from Insert and BatchInsert

File: cmd/go-ycsb/client.go [hunk 2]
```diff
@@ -43,6 +43,15 @@ func runClientCommandFunc(cmd *cobra.Command, args []string, doTransactions bool
 		if cmd.Flags().Changed("target") {
 			globalProps.Set(prop.Target, strconv.Itoa(targetArg))
 		}
+
+		if cmd.Flags().Changed("batchMode") {
```
Comment (siddontang) [inline cmd/go-ycsb/client.go]: remove this from the flags

File: cmd/go-ycsb/client.go [hunk 3]
```diff
@@ -88,14 +97,21 @@ func runTransCommandFunc(cmd *cobra.Command, args []string) {
...
 	m.Flags().StringSliceVarP(&propertyValues, "prop", "p", nil, "Specify a property value with name=value")
 	m.Flags().StringVar(&tableName, "table", "", "Use the table name instead of the default \""+prop.TableNameDefault+"\"")
-	m.Flags().IntVar(&threadsArg, "threads", 1, "execute using n threads - can also be specified as the \"threadcount\" property")
-	m.Flags().IntVar(&targetArg, "target", 0, "attempt to do n operations per second (default: unlimited) - can also be specified as the \"target\" property")
+	m.Flags().IntVar(&threadsArg, "threads", 1, "Execute using n threads - can also be specified as the \"threadcount\" property")
+	m.Flags().IntVar(&targetArg, "target", 0, "Attempt to do n operations per second (default: unlimited) - can also be specified as the \"target\" property")
+}
+
+func loadModeCommand(m *cobra.Command) {
+	m.Flags().BoolVar(&batchMode, "batchMode", false, "Use the batch mode in workload, just the tikv support the batch mode")
```
Comment (siddontang) [inline cmd/go-ycsb/client.go]: use properties, no flags

File: db/basic/db.go (hunk 1 above)
Comment (siddontang) [reply db/basic/db.go]: +1

File: pkg/client/client.go [hunk 4]
```diff
@@ -110,17 +114,25 @@ func (w *worker) run(ctx context.Context) {
...
+			}
 		}
 
 		if err != nil {
+			// TODO: add error log
 			break
 		}
 
-		w.opsDone++
+		w.opsDone = w.opsDone + int64(insertCount)
```
Comment (siddontang) [inline pkg/client/client.go]: w.opsDone += int64(insertCount)

File: pkg/prop/prop.go [hunk 5]
```diff
@@ -33,6 +33,11 @@ const (
 	DoTransactions     = "dotransactions"
 	Status             = "status"
 	Label              = "label"
+	// batch mode
+	DoBatch          = "batch.mode"
```
Comment (siddontang) [inline pkg/prop/prop.go]: no need batch mode, if batch size > 1, we can think it is in batch mode.

**Comment by liukun4515** [comment]
PTAL @Connor1996

**Comment by siddontang** [comment]
do you compare batch vs no batch insert?

**Comment by liukun4515** [comment]
```
recordcount=20000000
fieldcount=10
fieldlength=8
insertorder=hashed
threadcount=200
````
No batch insert:
```
INSERT - Takes(s): 200.0, Count: 10498845, QPS: 52501.0, Avg(us): 3776, Min(us): 140, Max(us): 502912, 95th(us): 10000, 99th(us): 13000
```


Batch insert:  **batch.size=16**
```
BATCH_INSERT - Takes(s): 60.0, Count: 840079, QPS: 14010.3, Avg(us): 6672, Min(us): 332, Max(us): 127928, 95th(us): 13000, 99th(us): 18000
```
image

@siddontang

**Comment by siddontang** [comment]
why batch qps is smaller than no batch

**Comment by liukun4515** [comment]
In the batch mode, the QPS of record inserted = batch_insert_QPS * batch.size.
The QPS of record inserted in batch mode is about 224K.
The QPS of record inserted in no batch is about 52K.

**Pull request merged**